hledger/Hledger/Cli/Commands/Balance.hs

```diff
@@ -319,12 +323,58 @@ balance opts@CliOpts{reportopts_=ropts} j = do
+          let budget = budgetJournal opts j
+              j' = budgetRollUp opts budget j
+              report = multiBalanceReport ropts (queryFromOpts d ropts) j'
+              budgetReport = multiBalanceReport ropts (queryFromOpts d ropts) budget
+              render = case format of
+                -- XXX: implement csv rendering
+                "csv" -> (++ "\n") . printCSV . multiBalanceReportAsCsv ropts
+                _     -> multiBalanceReportWithBudgetAsText ropts budgetReport
+          writeOutput opts $ render report
+          
+          | otherwise -> do
           let report = multiBalanceReport ropts (queryFromOpts d ropts) j
               render = case format of
-                "csv" -> \ropts r -> (++ "\n") $ printCSV $ multiBalanceReportAsCsv ropts r
-                _     -> multiBalanceReportAsText
-          writeOutput opts $ render ropts report
+                "csv" -> (++ "\n") . printCSV . multiBalanceReportAsCsv ropts
+                _     -> multiBalanceReportAsText ropts
+          writeOutput opts $ render report
+
+-- | Re-map account names to closet parent with periodic transaction from budget.
+-- Accounts that dont have suitable parent are either remapped to "<unbudgeted>:topAccount" 
+-- or left as-is if --show-unbudgeted is provided 
+budgetRollUp :: CliOpts -> Journal -> Journal -> Journal
+budgetRollUp CliOpts{rawopts_=rawopts} budget j = j { jtxns = remapTxn <$> jtxns j }
+    where
+        budgetAccounts = nub $ concatMap (map paccount . ptpostings) $ jperiodictxns budget
+        remapAccount origAcctName = remapAccount' origAcctName
+          where 
+            remapAccount' acctName
+              | acctName `elem` budgetAccounts = acctName
+              | otherwise = 
+                case parentAccountName acctName of
+                  "" | boolopt "show-unbudgeted" rawopts -> origAcctName
+                     | otherwise              -> T.append (T.pack "<unbudgeted>:") acctName
+                  parent -> remapAccount' parent
+        remapPosting p = p { paccount = remapAccount $ paccount p, porigin = Just . fromMaybe p $ porigin p }
+        remapTxn = mapPostings (map remapPosting)
+        mapPostings f t = txnTieKnot $ t { tpostings = f $ tpostings t }
+
+-- | Generate journal of all periodic transactions in the given journal for the
+-- entireity of its history or reporting period, whatever is smaller
+budgetJournal :: CliOpts -> Journal -> Journal
+budgetJournal opts j = journalBalanceTransactions' opts j { jtxns = budget }
+  where 
+    dates = spanIntersect (jdatespan j) (periodAsDateSpan $ period_ $ reportopts_ opts)
+    budget = [makeBudget t | pt <- jperiodictxns j, t <- runPeriodicTransaction pt dates]
+    makeBudget t = txnTieKnot $ t { tdescription = T.pack "Budget transaction" }
+    journalBalanceTransactions' opts j =
+      let assrt = not . ignore_assertions_ $ inputopts_ opts
+      in
+       either error' id $ journalBalanceTransactions assrt j
```

Comment:
Does `journalBalanceTransactions` still needed here? I would expect that at `balance` you already getting everything balanced. The reason why I've put it in `hledger-budget.hs` is because I was applying `runModifierTransaction` and I need to perform this befor bucketing.

Reply:
Sorry. I forgot that it produces completely independent `Journal` and thus requires all this post-processing applied.